statesync/syncer.go

```diff
@@ -0,0 +1,272 @@
+package statesync
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p"
+	"github.com/tendermint/tendermint/proxy"
+	sm "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tendermint/version"
+)
+
+const (
+	// chunkFetchers is the number of concurrent chunk fetchers to run.
+	chunkFetchers = 4
+	// chunkTimeout is the timeout before refetching a chunk, possibly with a different peer.
+	chunkTimeout = 10 * time.Second
+)
+
+// errNoSnapshots is returned by Sync() when no viable snapshots are found in the pool.
+var errNoSnapshots = errors.New("no viable snapshots found")
+
+// syncer runs a state sync against an ABCI app. Typical usage:
+//
+// 1) Create the syncer: newSyncer(...)
+// 2) Discover snapshots, and feed them via syncer.AddSnapshot()
+// 3) Start a sync via Sync(), which will request chunks from peers
+// 4) Concurrently feed chunks via syncer.AddChunk()
+// 5) Bootstrap node with the returned state and commit
+type syncer struct {
+	logger      log.Logger
+	stateSource StateSource
+	conn        proxy.AppConnSnapshot
+	connQuery   proxy.AppConnQuery
+	snapshots   *snapshotPool
+
+	mtx      sync.RWMutex
+	snapshot *snapshot
+	chunks   *chunkQueue
+}
+
+// newSyncer creates a new syncer.
+func newSyncer(logger log.Logger, conn proxy.AppConnSnapshot, connQuery proxy.AppConnQuery,
+	stateSource StateSource) *syncer {
+	return &syncer{
+		logger:      logger,
+		stateSource: stateSource,
+		conn:        conn,
+		connQuery:   connQuery,
+		snapshots:   newSnapshotPool(stateSource),
+		chunks:      nil,
+	}
+}
+
+// AddChunk adds a chunk to the chunk queue, if any. It returns false if the chunk has already
+// been added to the queue, or an error if there's no sync in progress. The height and format
+// are used as a sanity-check, in case the reactor receives unsolicited chunks for whatever reason.
+func (s *syncer) AddChunk(chunk *chunk) (bool, error) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	if s.snapshot == nil || s.chunks == nil {
```

Comment:
is it possible to have one, but not the other?

Reply:
It shouldn't be, but never hurts to be safe.